File: packages/desktop-releases/manifest.mjs
export const RELEASE_API = 'https://axiom.stream/api/releases/latest';
export const VERSION_PATTERN = /^(0|[1-9]\d{0,8})\.(0|[1-9]\d{0,8})\.(0|[1-9]\d{0,8})$/;
export const FORMATS = { mac: ['pkg'], win: ['exe'], linux: ['AppImage', 'deb', 'pacman'] };
export const CLI_FORMATS = { mac: ['pkg'], win: ['exe'], linux: ['sh'] };
export const GITHUB_RELEASES = 'https://github.com/astrea-foundation/axiom-releases/releases';

export function compareVersions(a, b) {
  if (!VERSION_PATTERN.test(a) || !VERSION_PATTERN.test(b)) throw new Error('Invalid release version');
  const left = a.split('.').map(Number), right = b.split('.').map(Number);
  for (let i = 0; i < 3; i++) if (left[i] !== right[i]) return Math.sign(left[i] - right[i]);
  return 0;
}
export function artifactFormat(name) {
  if (/\.pkg\.tar\.(xz|zst)$/.test(name)) return 'pacman';
  return /\.(AppImage|deb|dmg|pkg|zip|exe|sh)$/.exec(name)?.[1] ?? null;
}
// Recursively sorted object keys, compact UTF-8 JSON, no trailing newline.
export function canonicalJson(value) {
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(',')}]`;
  if (value && typeof value === 'object') return `{${Object.keys(value).sort().map(key => `${JSON.stringify(key)}:${canonicalJson(value[key])}`).join(',')}}`;
  return JSON.stringify(value);
}
export function unsignedManifest(value) { const {signature, ...manifest} = value; return manifest; }
export function parseRelease(value) {
  if (!value || typeof value !== 'object' || Array.isArray(value) || typeof value.version !== 'string' || typeof value.revision !== 'string' || ![2, 3].includes(value.schemaVersion) || !VERSION_PATTERN.test(value.version ?? '')
    || !/^[a-f0-9]{40}$/.test(value.revision ?? '') || !['signed', 'unsigned'].includes(value.signing)
    || !Number.isSafeInteger(value.sequence) || value.sequence < 1
    || !Array.isArray(value.downloads) || !Array.isArray(value.cliDownloads)
    || !value.downloads.length || value.downloads.length + value.cliDownloads.length > 100) throw new Error('Invalid release manifest');
  const keys = ['schemaVersion', 'version', 'revision', 'sequence', 'signing', 'downloads', 'cliDownloads', 'signature'];
  if (Object.keys(value).some(key => !keys.includes(key))) throw new Error('Unknown release field');
  if (value.signing === 'signed' && (!value.signature || typeof value.signature.keyId !== 'string' || typeof value.signature.value !== 'string' || !/^[a-f0-9]{16}$/.test(value.signature.keyId ?? '')
    || !/^[A-Za-z0-9+/]{85}[AQgw]==$/.test(value.signature.value ?? '') || Object.keys(value.signature).sort().join() !== 'keyId,value')) throw new Error('Missing release signature');
  if (value.signing === 'unsigned' && value.signature !== undefined) throw new Error('Unsigned release has a signature');
  const seen = new Set();
  for (const [product, files, formats] of [['desktop', value.downloads, FORMATS], ['cli', value.cliDownloads, CLI_FORMATS]]) {
    for (const file of files) {
      if (!file || typeof file !== 'object' || Array.isArray(file) || ['name','product','platform','arch','format','sha256','url','githubUrl'].some(key => typeof file[key] !== 'string')) throw new Error('Invalid release artifact');
      const format = artifactFormat(file.name ?? '');
      const namedArch = /-universal[.-]/.test(file.name) ? 'universal' : /-arm64[.-]/.test(file.name) ? 'arm64' : /-(x64|x86_64|amd64)[.-]/.test(file.name) ? 'x64' : null;
      const key = `${product}:${file.platform}:${file.arch}:${format}`;
      if (Object.keys(file).sort().join() !== 'arch,bytes,format,githubUrl,name,platform,product,sha256,url'
        || file.product !== product || !/^[A-Za-z0-9][A-Za-z0-9._-]{0,199}$/.test(file.name ?? '')
        || !file.name.includes(`-${value.version}-`) || !(value.schemaVersion === 3 ? (file.platform === 'linux' ? ['x64'] : ['universal']) : ['x64', 'arm64']).includes(file.arch)
        || !Object.hasOwn(formats, file.platform) || !formats[file.platform].includes(format)
        || file.format !== format || namedArch !== file.arch || seen.has(key)
        || !Number.isSafeInteger(file.bytes) || file.bytes < 1 || file.bytes > 2 * 1024 ** 3
        || !/^[a-f0-9]{64}$/.test(file.sha256 ?? '')
        || file.url !== `https://cdn.axiom.stream/downloads/${value.version}/${file.name}`
        || file.githubUrl !== `${GITHUB_RELEASES}/download/v${value.version}/${file.name}`) throw new Error('Invalid release artifact');
      seen.add(key);
    }
  }
  return structuredClone(value);
}

export function requireCompleteRelease(value) {
  const release = parseRelease(value);
  const expected = new Set([
    ...['desktop','cli'].flatMap(product => ['mac','win'].flatMap(platform => (release.schemaVersion === 3 ? ['universal'] : ['x64','arm64']).map(arch=>`${product}:${platform}:${arch}:${platform === 'mac' ? 'pkg' : 'exe'}`))),
    ...['AppImage','deb','pacman'].map(format=>`desktop:linux:x64:${format}`), 'cli:linux:x64:sh',
  ]);
  for (const file of [...release.downloads,...release.cliDownloads]) {
    if (!expected.delete(`${file.product}:${file.platform}:${file.arch}:${file.format}`)) throw new Error('Unexpected release target');
  }
  if (expected.size) throw new Error(`Incomplete release: ${[...expected].join(', ')}`);
  return release;
}

File: apps/axiomcli/src/updates/manifest.rs
use std::collections::HashSet;

use anyhow::{Context as _, ensure};
use base64::{Engine as _, engine::general_purpose::STANDARD};
use ed25519_dalek::{Signature, VerifyingKey};
use serde::{Deserialize, Serialize};
use sha2::{Digest as _, Sha256};

pub const MAX_FEED_BYTES: usize = 128 * 1024;
pub const MAX_ARTIFACT_BYTES: u64 = 2 * 1024 * 1024 * 1024;

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Artifact {
    pub name: String,
    pub product: String,
    pub platform: String,
    pub arch: String,
    pub format: String,
    pub bytes: u64,
    pub sha256: String,
    pub url: String,
    pub github_url: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ReleaseSignature {
    pub key_id: String,
    pub value: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Release {
    pub schema_version: u32,
    pub version: String,
    pub revision: String,
    pub sequence: u64,
    pub signing: String,
    pub downloads: Vec<Artifact>,
    pub cli_downloads: Vec<Artifact>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<ReleaseSignature>,
}

pub fn version_parts(version: &str) -> anyhow::Result<[u32; 3]> {
    let parts = version.split('.').collect::<Vec<_>>();
    ensure!(parts.len() == 3, "Invalid release version");
    let mut result = [0; 3];
    for (i, part) in parts.iter().enumerate() {
        ensure!(
            !part.is_empty()
                && part.len() <= 9
                && part.bytes().all(|c| c.is_ascii_digit())
                && (part.len() == 1 || !part.starts_with('0')),
            "Invalid release version"
        );
        result[i] = part.parse()?;
    }
    Ok(result)
}

fn lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|c| c.is_ascii_digit() || (b'a'..=b'f').contains(&c))
}

pub fn parse_release(bytes: &[u8]) -> anyhow::Result<Release> {
    ensure!(bytes.len() <= MAX_FEED_BYTES, "Release feed is too large");
    let release: Release = serde_json::from_slice(bytes)?;
    ensure!(
        matches!(release.schema_version, 2 | 3)
            && release.sequence > 0
            && release.sequence <= 9_007_199_254_740_991
            && lower_hex(&release.revision, 40)
            && matches!(release.signing.as_str(), "signed" | "unsigned")
            && !release.downloads.is_empty()
            && release.downloads.len() + release.cli_downloads.len() <= 100,
        "Invalid release manifest"
    );
    version_parts(&release.version)?;
    match (&release.signature, release.signing.as_str()) {
        (Some(signature), "signed") => {
            ensure!(
                lower_hex(&signature.key_id, 16) && STANDARD.decode(&signature.value)?.len() == 64,
                "Invalid release signature"
            );
        }
        (None, "unsigned") => {}
        _ => anyhow::bail!("Missing release signature"),
    }
    let mut seen = HashSet::new();
    for (product, files) in [
        ("desktop", &release.downloads),
        ("cli", &release.cli_downloads),
    ] {
        for file in files {
            let allowed: &[&str] = match (product, file.platform.as_str()) {
                (_, "mac") => &["pkg"],
                (_, "win") => &["exe"],
                ("desktop", "linux") => &["AppImage", "deb", "pacman"],
                ("cli", "linux") => &["sh"],
                _ => &[],
            };
            let format =
                if file.name.ends_with(".pkg.tar.xz") || file.name.ends_with(".pkg.tar.zst") {
                    "pacman"
                } else {
                    file.name.rsplit('.').next().unwrap_or_default()
                };
            let marker = |a: &str| {
                file.name.contains(&format!("-{a}.")) || file.name.contains(&format!("-{a}-"))
            };
            let arch = if marker("universal") {
                "universal"
            } else if marker("arm64") {
                "arm64"
            } else if ["x64", "x86_64", "amd64"].iter().any(|a| marker(a)) {
                "x64"
            } else {
                ""
            };
            ensure!(
                file.product == product
                    && file.name.len() <= 200
                    && file
                        .name
                        .as_bytes()
                        .first()
                        .is_some_and(u8::is_ascii_alphanumeric)
                    && file
                        .name
                        .bytes()
                        .all(|c| c.is_ascii_alphanumeric() || b"._-".contains(&c))
                    && file.name.contains(&format!("-{}-", release.version))
                    && match (release.schema_version, file.platform.as_str()) {
                        (3, "linux") => file.arch == "x64",
                        (3, _) => file.arch == "universal",
                        _ => matches!(file.arch.as_str(), "x64" | "arm64"),
                    }
                    && arch == file.arch
                    && allowed.contains(&format)
                    && file.format == format
                    && file.bytes > 0
                    && file.bytes <= MAX_ARTIFACT_BYTES
                    && lower_hex(&file.sha256, 64)
                    && file.url
                        == format!(
                            "https://cdn.axiom.stream/downloads/{}/{}",
                            release.version, file.name
                        )
                    && file.github_url
                        == format!(
                            "https://github.com/astrea-foundation/axiom-releases/releases/download/v{}/{}",
                            release.version, file.name
                        )
                    && seen.insert((product, &file.platform, &file.arch, &file.format)),
                "Invalid release artifact"
            );
        }
    }
    Ok(release)
}

pub fn verify_release(release: &Release, trusted_keys: &str) -> anyhow::Result<()> {
    ensure!(
        release.signing == "signed",
        "Automatic updates require a signed release"
    );
    let signature = release
        .signature
        .as_ref()
        .context("Missing release signature")?;
    let key = trusted_keys
        .split(',')
        .filter_map(|hex| hex::decode(hex.trim()).ok())
        .find(|key| {
            key.len() == 32 && hex::encode(Sha256::digest(key)).starts_with(&signature.key_id)
        })
        .context("This build does not trust the update signing key")?;
    let key: [u8; 32] = key
        .try_into()
        .map_err(|_| anyhow::anyhow!("Invalid update key"))?;
    let mut value = serde_json::to_value(release)?;
    value
        .as_object_mut()
        .context("Invalid release")?
        .remove("signature");
    let payload = canonical_json(&value)?;
    let signature = Signature::from_slice(&STANDARD.decode(&signature.value)?)?;
    VerifyingKey::from_bytes(&key)?
        .verify_strict(&payload, &signature)
        .context("Invalid update signature")
}

pub(super) fn canonical_json(value: &serde_json::Value) -> anyhow::Result<Vec<u8>> {
    fn write(value: &serde_json::Value, out: &mut Vec<u8>) -> anyhow::Result<()> {
        match value {
            serde_json::Value::Object(map) => {
                out.push(b'{');
                let mut keys: Vec<_> = map.keys().collect();
                keys.sort();
                for (i, key) in keys.into_iter().enumerate() {
                    if i > 0 {
                        out.push(b',');
                    }
                    serde_json::to_writer(&mut *out, key)?;
                    out.push(b':');
                    write(&map[key], out)?;
                }
                out.push(b'}');
            }
            serde_json::Value::Array(items) => {
                out.push(b'[');
                for (i, item) in items.iter().enumerate() {
                    if i > 0 {
                        out.push(b',');
                    }
                    write(item, out)?;
                }
                out.push(b']');
            }
            _ => serde_json::to_writer(out, value)?,
        }
        Ok(())
    }
    let mut out = Vec::new();
    write(value, &mut out)?;
    Ok(out)
}
